src/core/mtu_discovery.c

```diff
@@ -0,0 +1,237 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    This module handles the MTU discovery logic.
+
+    Upon a new path being validated, MTU discovery is started on that path.
+    This is done by sending a probe packet larger than the current MTU.
+
+    If the probe packet is acknowledged, that is set as the current MTU and a
+    new probe packet is sent. This is repeated until the maximum allowed MTU is
+    reached.
+
+    If a probe packet is not ACKed, the probe at the same size will be retried.
+    If this fails QUIC_DPLPMTUD_MAX_PROBES times, max MTU is considered found
+    and searching stops.
+
+    Once searching has stopped, discovery will stay idle until
+    QUIC_DPLPMTUD_RAISE_TIMER_TIMEOUT has passed. The next send will then
+    trigger a new MTU discovery period, unless maximum allowed MTU is already
+    reached.
+
+    The current algorithm is very simplistic, increasing by 80 bytes each
+    probe. A special case is added so 1500 is always a checked value, as 1500
+    is often the max allowed over the internet.
+
+--*/
+
+#include "precomp.h"
+#ifdef QUIC_CLOG
+#include "mtu_discovery.c.clog.h"
+#endif
+
+CXPLAT_STATIC_ASSERT(CXPLAT_MAX_MTU >= QUIC_DPLPMUTD_DEFAULT_MAX_MTU, L"Default max must not be more than max");
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+static
+void
+QuicMtuDiscoverySendProbePacket(
+    _In_ QUIC_MTU_DISCOVERY* MtuDiscovery
+    )
+{
+    QUIC_CONNECTION* Connection =
+        CXPLAT_CONTAINING_RECORD(MtuDiscovery, QUIC_CONNECTION, MtuDiscovery);
+    QuicSendSetSendFlag(&Connection->Send, QUIC_CONN_SEND_FLAG_DPLPMTUD);
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+static
+void
+QuicMtuDiscoveryMoveToSearchComplete(
+    _In_ QUIC_MTU_DISCOVERY* MtuDiscovery
+    )
+{
+    QUIC_CONNECTION* Connection =
+        CXPLAT_CONTAINING_RECORD(MtuDiscovery, QUIC_CONNECTION, MtuDiscovery);
+    MtuDiscovery->IsSearching = FALSE;
+    MtuDiscovery->SearchCompleteEnterTimeUs = CxPlatTimeUs64();
+    QuicTraceLogConnInfo(
+        MtuSearchComplete,
+        Connection,
+        "Mtu Discovery Entering Search Complete at MTU %u",
+        MtuDiscovery->CurrentMtu);
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+static
+uint16_t
+QuicGetNextProbeSize(
+    _In_ const QUIC_MTU_DISCOVERY* MtuDiscovery
+    )
+{
+    //
+    // N.B. This algorithm must always be increasing. Other logic in the module
+    // depends on that behavior.
+    //
+
+    uint16_t Mtu = MtuDiscovery->CurrentMtu + QUIC_DPLPMTUD_INCREMENT;
+    if (Mtu > MtuDiscovery->MaxMtu) {
+        Mtu = MtuDiscovery->MaxMtu;
+    } else if (Mtu == 1520) {
```

Comment:
This doesn't work if the INCREMENT changes, right? How can we make this more generic? Maybe if the increment skips 1500, then go back to 1500 and increment from there?